Post for Engineering Principles (#48)
raft-tech/raft-tech.github.io: _posts/2020-02-19-engineering-principles.md

---
layout: post
title:  "Raft Engineering Principles"
date:   2020-02-19
categories:
featimg: /assets/images/featimage/engineering_principles.jpg
bgimg: /assets/images/bgimg/eng.jpg
---

These principles are a part of our culture and each project we undertake. They are not specific to any programming language, framework, platform, or tool, but specific to the fundamentals and best practices we use when building and maintaining software for ourselves and our customers. By being open and transparent about sharing these principles, we want to show how the sausage is made.

## Tl;DR

- Start With the User 
- Find the Right Tools for the Right Job 
- Follow an Opinionated Gitflow 
- Sustain High Quality (Agile)
- Stand on Shoulders of Giants
- Choose Open Source Over Vendor Lock-In
- Cut the Buzz Words
- Design for Scale
- Always Be Learning
- Strive for Automation
- Share Lessons Learned
- Obey the 15-Minute Rule
- Code Readability Over Short Code
- Learn the Tools

## Start With the User 

> Every behind-the-scenes decision we take as engineers has an impact on the end user. The impact might be direct (e.g. adding a large dependency), indirect (e.g. fixing a CVE), or specific to a subset of users. But thinking about that impact during every stage of the decision-making process allows us to pivot our approach and cross collaborate with our user experience team. 

Reviewer: But thinking about that impact during every stage of the decision-making process allows us to collaborate with our user-experience team and, when advantageous, pivot our approach.

Author: Implemented

## Find the Right Tools for the Right Job 

> We have a generalist approach when it comes to programming languages, frameworks, libraries, platforms, etc. We believe the right tools can only be used for the right job when engineers truly understand the mission. Once we understand the problem that the solution is being built for, we can choose/advise on the best tools. Problems, like data validation, are easier to solve with strongly typed languages with good type safety (e.g. Scala, Java) than dynamically typed languages (e.g. Ruby, Python). Running application that need horizontal scalability won't work well as well on bare bones EC2 as containerizing it on Kubernetes. A problem requiring a machine learning model will operate better on R or Python ecosystems than most other languages. Below are some questions we ask ourselves when choosing these tools: 

Reviewer: We have a generalist approach when it comes to programming languages, frameworks, libraries, platforms, etc. We believe that understanding the mission by thoroughly evaluating the problem to be solved is a precursor to selecting the most effective tool chain. Problems like data validation are easier to solve with strongly typed languages (e.g. Scala, Java) rather than dynamically typed (e.g. Ruby, Python). Running an application that needs horizontal scalability on bare bones EC2 won't work as well as containerizing it on Kubernetes. A problem requiring a machine learning model will operate better in an R or Python ecosystem than in most others. Below are some questions we ask ourselves when choosing these tools:

- What are the ecosystems and communities around these tools?
- Will the tools allow us to pivot, iterate, and ship faster?
- Will the tools be around 5 to 10 years from now?
- What is the learning curve for the new members?
- Does the problem we're trying to fix have commonalities with projects we've done in the past? 

## Follow an Opinionated GitFlow 

Once you understand the fundamentals of a Git repository, it is clear why Git is the preferred version control tool. However, learning Git internals takes time and patience. We've seen senior engineers struggle with Git problems that sidetrack them from adding business value. That is why we follow an opinionated GitFlow with aliases, origins, upstreams, separate branches, pull requests, and more. Using an opinionated GitFlow ensures that our engineering team is always on the same page.

## Sustain High Quality (Agile)

> We don't like to pigeonhole our team into following a specific software development methodology as long as it is [not SAFe](https://software.af.mil/wp-content/uploads/2019/12/CSO-MFR-on-Agile-Frameworks-12282019.pdf). Rather, the cornerstones of our software development principles allow our engineering teams to produce high-quality features quickly.

Reviewer: ```suggestion
Our teams are not pigeonholed into following a specific software development methodology, as long as it is[not SAFe](https://software.af.mil/wp-content/uploads/2019/12/CSO-MFR-on-Agile-Frameworks-12282019.pdf). Rather, our software development principles allow our engineering teams to produce high-quality features, quickly.
```

Reviewer 2: I think the emphasis is that **we** as raft do not pigeonhole our teams. Sometimes teams get pigeonholed.

Reviewer: That's a good point.

> - Sustainable high quality

Reviewer: ```suggestion

```

>Buying an app will not ensure the outcomes you want any more than buying a new pair of running shoes will ensure that you finish a marathon. You still have to do the work — every piece of it, top to bottom. 

All engineers are encouraged to contribute their learnings back to the open source by either closing issues, sending pull requests, or updating documentation. 

## Cut the Buzz Words

> We're not fans of buzz words. Buzz words are great for helping creative sales teams simplify complicated engineering topics for non-technical folk. However, more often than not, they are misleading and favor vendor preference. For example, this blog is completely "Serverless," but we have no idea whether it runs on AWS, Azure, Bare Metals, etc. We don't throw "Machine Learning" at "Big Data" and expect magical insights. Rather, we focus on the business insights our users desire and work backwards to pick the learning algorithms that will work best over time with the least [Mean Squared Error](https://en.wikipedia.org/wiki/Mean_squared_error).

Reviewer: ```suggestion
We're not fans of buzz words. Buzz words are great for helping creative sales teams simplify complicated engineering topics for non-technical folk. However, more often than not, they are misleading and exhibit a vendor bias. For example, this blog is completely "Serverless," but we have no idea whether it runs on AWS, Azure, Bare Metals, etc. We don't throw "Machine Learning" at "Big Data" and expect magical insights. Rather, we focus on the business insights our users desire and work backwards to pick the learning algorithms that will work best over time with the least [Mean Squared Error](https://en.wikipedia.org/wiki/Mean_squared_error).
```

> We're not fans of buzz words. Buzz words are great for helping creative sales teams simplify complicated engineering topics for non-technical folk. However, more often than not, they are misleading and favor vendor preference. For example, this blog is completely "Serverless," but we have no idea whether it runs on AWS, Azure, Bare Metals, etc. We don't throw "Machine Learning" at "Big Data" and expect magical insights. Rather, we focus on the business insights our users desire and work backwards to pick the learning algorithms that will work best over time with the least [Mean Squared Error](https://en.wikipedia.org/wiki/Mean_squared_error).

Reviewer: However, more often than not, they are misleading and exhibit a vendor bias.

## Design for Scale 

> We work on public facing applications that parse, validate, and ingest over 40 million records in 60 days (with over 80% of the submissions in the last 10 days). While designing applications, we think about how the applications will scale early on in the design process. Throwing cloud resources at a scale problem only gets you so far. We choose programming languages capable of handling concurrency with ease (e.g. Scala), software patterns like Command Query Responsibility Segregation (CQRS), modular micro-services, event driven systems (e.g. Kafka), and platforms capable of keeping up with scale (e.g. Kubernetes). These engineering decisions are critical to the success of the applications we work on in the public sector, so we dedicate specific sprints early in the project life cycle to make these decisions. The decisions we make aren't set in stone but rather allow for flexibility and being able to pivot down the line without significant cost. 

Reviewer: These choices are critical to the success of the applications we work on in the public sector, so we dedicate specific sprints early in the project life cycle to make these engineering decisions.

## Strive for Automation

Any manually-performed, technical task that _can_ be automated _must_ be automated. Automation provides us repeatability, time to focus on value-driven development, assured security, and shorter feedback loops when troubleshooting issues. Each pull request merged into master must go through an automated execution of unit tests, integration tests, code coverage, static code scanning, dependency checking, building, and deploying the container to each development platform. Each image built must live on a private registry which is automatically scanned for high and critical CVEs. Any cron job must be a K8s CronJob, rather than hidden behind SSH. Backups and restores must be automated, recur at regular intervals, and only alert us when they fail. End-to-end integration testing of our APIs using Newman must be automated and alert us when integration tests fail. 

## Share Lessons Learned 

> Engineering, especially software engineering, allows for starting small, validating, learning from mistakes, and iterating. Sharing the knowledge gained doesn't go far enough if there isn't a formal process of incorporating that knowledge into a document, script, or automated pipeline. At the end of each retrospective, we hold a "dev talk" where each engineer has a chance to share what they learned and how they've incorporated it into our current processes. 

Reviewer: Knowledge gained doesn't go far unless there is a formal process of incorporating that knowledge into a document, script, or automated pipeline.
